src/main/kotlin/com/openlattice/transporter/TransporterUtils.kt

```diff
@@ -0,0 +1,110 @@
+            val rs = st.executeQuery()
+            rs.next()
+        }
+    }
+}
+
+val pkCols = listOf(PostgresColumn.ENTITY_SET_ID, PostgresColumn.ID)
+val pk = pkCols.joinToString(", ") { it.name }
+
+/**
+ * column bindings are
+ * 1 - partitions array
+ * 2 - entity set ids array
+ *
+ */
+fun updateOneBatchForProperty(destTable: String, propCol: Map.Entry<UUID, TransporterColumn>): String {
+    val srcCol = propCol.value.srcCol
+    val destCol = propCol.value.destColName
+    val id = propCol.key
+
+    val updateLastPropagate = "UPDATE ${PostgresTable.DATA.name} " +
+            "SET ${PostgresColumn.LAST_PROPAGATE.name} = ${DataTables.LAST_WRITE.name} " +
+            "WHERE ${PostgresColumn.PARTITION.name} = ANY(?) " +
+            " AND ${PostgresColumn.ENTITY_SET_ID.name} = ANY(?) " +
+            " AND ${PostgresColumn.PROPERTY_TYPE_ID.name} = '${id}' " +
+            " AND ${PostgresColumn.ORIGIN_ID.name} = '${IdConstants.EMPTY_ORIGIN_ID.id}' " +
+            " AND ${DataTables.LAST_WRITE.name} > ${PostgresColumn.LAST_PROPAGATE.name} " +
+            "RETURNING ${PostgresColumn.ENTITY_SET_ID.name}, ${PostgresColumn.ID.name}, ${PostgresColumn.VERSION.name}, " +
+            " CASE WHEN ${PostgresColumn.VERSION.name} > 0 then $srcCol else null end as $destCol"
+
+    // this is almost certainly not necessary but if ids is already synced in the same transaction
+    val createMissingRows = "INSERT INTO $destTable " +
+            "($pk) " +
+            "SELECT $pk " +
+            "FROM src " +
+            "WHERE ${PostgresColumn.VERSION.name} > 0 " +
+            "ON CONFLICT ($pk) DO NOTHING"
+    val modifyDestination = "UPDATE $destTable t " +
+            "SET $destCol = src.$destCol " +
+            "FROM src " +
+            "WHERE " + pkCols.joinToString(" AND ") { "t." + it.name + " = src." + it.name}
+
+    return "WITH src as (${updateLastPropagate}), inserted as ($createMissingRows) $modifyDestination"
+}
+
+/**
+ * column bindings are
+ * 1 - partitions array
+ * 2 - entity set ids array
+ *
+ */
+fun updateIdsForEntitySets(destTable: String): String {
+    val updateLastPropagate = "UPDATE ${PostgresTable.IDS.name} " +
+            "SET ${PostgresColumn.LAST_PROPAGATE.name} = ${DataTables.LAST_WRITE.name} " +
+            "WHERE ${PostgresColumn.PARTITION.name} = ANY(?) " +
+            " AND ${PostgresColumn.ENTITY_SET_ID.name} = ANY(?) " +
+            " AND ${DataTables.LAST_WRITE.name} > ${PostgresColumn.LAST_PROPAGATE.name} " +
+            "RETURNING ${PostgresColumn.ENTITY_SET_ID.name}, ${PostgresColumn.ID.name}, ${PostgresColumn.VERSION.name}"
+    val createMissingRows = "INSERT INTO $destTable ($pk) " +
+            "SELECT $pk " +
+            "FROM src " +
+            "WHERE ${PostgresColumn.VERSION.name} > 0 " +
+            "ON CONFLICT ($pk) DO NOTHING"
+    val deleteRows = "DELETE FROM $destTable t " +
+            "USING src " +
+            "WHERE src.${PostgresColumn.VERSION.name} <= 0 AND " +
```

Comment:
> I just pushed a change to add hard deletes.
>

This will I think only handle hard deletes , if they are not yet picked up by the background task (that picks up rows where version = 0), because after that they will be permanently deleted. Soft deletes should be fine I think this way.

Reply:
Oh, or will that task wait until last_propagate (or last_materialize..) is set before deleting entities?

Reply:
Yeah we'll need to deal with that in BackgroundIndexedEntitiesDeletionService, but I'm waiting on that PR until this all shakes out. :)